localstack/services/dynamodb/provider.py

```diff
@@ -0,0 +1,1351 @@
+    def scan(self, context: RequestContext, scan_input: ScanInput) -> ScanOutput:
+        return self.forward_request(context)
+
+    @handler("BatchWriteItem", expand=False)
+    def batch_write_item(
+        self,
+        context: RequestContext,
+        batch_write_item_input: BatchWriteItemInput,
+    ) -> BatchWriteItemOutput:
+        existing_items = []
+        unprocessed_put_items = []
+        unprocessed_delete_items = []
+        request_items = batch_write_item_input["RequestItems"]
+        for table_name in sorted(request_items.keys()):
+            for request in request_items[table_name]:
+                for key in ["PutRequest", "DeleteRequest"]:
+                    inner_request = request.get(key)
+                    if inner_request:
+                        if self.should_throttle("BatchWriteItem"):
+                            if key == "PutRequest":
+                                unprocessed_put_items.append(inner_request)
+                            elif key == "DeleteRequest":
+                                unprocessed_delete_items.append(inner_request)
+                        else:
+                            item = ItemFinder.find_existing_item(inner_request, table_name)
+                            existing_items.append(item)
+
+        # forward request to backend
+        result = self.forward_request(context)
+
+        # determine and forward stream records
+        request_items = batch_write_item_input["RequestItems"]
+        records, unprocessed_items = self.prepare_batch_write_item_records(
+            request_items=request_items,
+            unprocessed_put_items=unprocessed_put_items,
+            unprocessed_delete_items=unprocessed_delete_items,
+            existing_items=existing_items,
+        )
+        streams_enabled_cache = {}
+        event_sources_or_streams_enabled = False
+        for record in records:
+            event_sources_or_streams_enabled = (
+                event_sources_or_streams_enabled
+                or has_event_sources_or_streams_enabled(
+                    record["eventSourceARN"], streams_enabled_cache
+                )
+            )
+        if event_sources_or_streams_enabled:
+            self.forward_stream_records(records)
+
+        # update response
+        if any(unprocessed_items):
+            table_name = list(request_items.keys())[0]
+            unprocessed = result["UnprocessedItems"]
+            if table_name not in unprocessed:
+                unprocessed[table_name] = []
+            for key in ["PutRequest", "DeleteRequest"]:
+                if any(unprocessed_items[key]):
+                    unprocessed_items[table_name].append({key: unprocessed_items[key]})
+            for key in list(unprocessed.keys()):
+                if not unprocessed.get(key):
+                    del unprocessed[key]
+
+            # TODO update CRC32 headers?
+            # fix_headers_for_updated_response(response)
```

Comment:
nit: I guess this TODO isn't necessary anymore, since this is tackled globally in the listener.